Bugfix 67214 minion swarm

dnessett opened the pull request:
### What does this PR do?
Fixes bugs in tests/minionswarm.py that were preventing it to run.

### What issues does this PR fix or reference?
Fixes #67214

### Previous Behavior
minionswarm failed to run

### New Behavior
minionswarm now runs correctly

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [ ] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

<!-- Please review Salt's Contributing Guide for best practices and guidance in
choosing the right branch:
https://docs.saltproject.io/en/master/topics/development/contributing.html -->

<!-- Additional guidance for pull requests can be found here:
https://docs.saltproject.io/en/master/topics/development/pull_requests.html -->

<!-- See GitHub's page on GPG signing for more information about signing commits
with GPG:
https://help.github.com/articles/signing-commits-using-gpg/ -->

Commit:
Fix bugs in minionswarm. Updated help with significant enhancements. Added username to salt-master command. Added ability to turn off open_mode.

Commit:
cleaned up some problems with the parser.add_options and added a
clarificationn to the config-dir help text.

Commit:
added changelog file

Commit:
Eliminated some trailing whitespace errors revealed by nox -e lint-tests

dwoz commented on tests/support/runtests.py:
```diff
@@ -50,7 +50,7 @@
 import salt.utils.path
 import salt.utils.platform
-import tests.support.paths as paths
+import support.paths as paths
```
Why is this needed?

dnessett replied:
The original threw an unfound error. runtests.py exists within salt/tests/support. Relative to runtests.py paths is not located at tests.support.paths, but rather at support.paths. This may have resulted from a reorganization of the directory structure sometime after the 2015 release in which minionswarm.py and runtests.py were moved (that's a guess).

dnessett replied:
I just spent some time reading about relative pathnames in the import statement. I used "import support.paths" simply because it worked. However, after reading, I'm not sure why it works. Perhaps "import paths" would be better, since paths.py exists in the same directory as runtests.py.

twangboy commented:
Looks like there are some pre-commit failures.

dnessett commented:
> Looks like there are some pre-commit failures.

OK. I will take a look at it. It may take some time since I am reorganizing my development environment.

dnessett commented:
I am very new to salt development, so I very possibly am missing something, I ran nox -e lint-tests in bugfix-67214, which is what is pushed to https://github.com/dnessett/salt and what is the basis of the PR. The results are:

```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ nox -e lint-tests
nox > Running session lint-tests
nox > Creating virtual environment (virtualenv) using python3 in .nox/lint-tests
nox > python -m pip install --progress-bar=off -U setuptools pip wheel
nox > python -m pip install --progress-bar=off -r requirements/static/ci/py3.10/linux.txt -r requirements/static/ci/py3.10/lint.txt
nox > pylint --rcfile=.pylintrc --disable=I tests/

------------------------------------
Your code has been rated at 10.00/10

nox > Session lint-tests was successful.
```
To check I was in the correct branch, I ran:
```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ git log
commit aeb11b3e615066804d9debd581b8663234a4bffc (HEAD -> bugfix-67214, origin/bugfix-67214)
Author: dnessett <dnessett@yahoo.cm>
Date:   Sat Mar 1 13:24:48 2025 -0700

    Eliminated some trailing whitespace errors revealed by nox -e lint-tests

commit ed91f4ceecfb92f773096951dd8a2ac4ee1f73a8
Author: dnessett <dnessett@yahoo.cm>
Date:   Sat Mar 1 12:24:11 2025 -0700

    added changelog file

commit 05f5e132d2d90844df3fbc432802a864c42c8dc2
Author: dnessett <dnessett@yahoo.cm>
Date:   Fri Feb 28 13:19:32 2025 -0700

    cleaned up some problems with the parser.add_options and added a
    clarificationn to the config-dir help text.

commit 85afd494938c53934e941eb7cb658a9cbcd42062
Author: dnessett <dnessett@yahoo.cm>
Date:   Thu Feb 27 11:31:06 2025 -0700

    Fix bugs in minionswarm. Updated help with significant enhancements. Added username to salt-master command. Added ability to turn off open_mode.

commit 2530c8f1d5b1440b70860d4dea453b6657bdf92b
Author: Daniel A. Wozniak <daniel.wozniak@broadcom.com>
Date:   Wed May 7 01:33:13 2025 -0700

    Do not fail when cached grains no longer exit
```
and:
```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ git branch
* bugfix-67214
  master
```
I then checked the changelog directory in case the changelog file was faulty, but it seems not:

```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt/changelog$ cat 67214.fixed.md 
Fixed bugs in minionswarm and added ability turn off open-mode.
```
So, I am stuck finding out what is causing pre-commit to fail.

lkubb commented:
`lint-tests`/`lint-salt` run `pylint` only. The pre-commit hooks that are failing are `black`/`isort` though. You'd run them via

`pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py`
`pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py`

In general, you should consider installing the `pre-commit` hook (the git hook) for `pre-commit` (the program) via `pre-commit install` (or `python -m pre_commit install`, if you're in a venv and want to use the pre-commit that's installed inside the venv). This would ensure you cannot commit before all hooks are fixed, thus avoiding the failing check in CI entirely.

dnessett commented:
> `lint-tests`/`lint-salt` run `pylint` only. The pre-commit hooks that are failing are `black`/`isort` though. You'd run them via
> 
> `pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py` `pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py`
> 
> In general, you should consider installing the `pre-commit` hook (the git hook) in your local `salt` checkout for `pre-commit` (the program) via `pre-commit install` (or `python -m pre_commit install`, if you're in a venv and want to ensure you're using the pre-commit that's installed inside the venv). This would ensure you cannot commit before all hooks are fixed, thus avoiding the failing check in CI entirely.

Thanks.

dnessett commented:
> > `lint-tests`/`lint-salt` run `pylint` only. The pre-commit hooks that are failing are `black`/`isort` though. You'd run them via
> > `pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py` `pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py`
> > In general, you should consider installing the `pre-commit` hook (the git hook) in your local `salt` checkout for `pre-commit` (the program) via `pre-commit install` (or `python -m pre_commit install`, if you're in a venv and want to ensure you're using the pre-commit that's installed inside the venv). This would ensure you cannot commit before all hooks are fixed, thus avoiding the failing check in CI entirely.
> pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py
> Thanks.

Sorry for the delay, I had some emergencies to resolve in another area of my volunteer work.

When I run:  `pre-commit run black/isort`
The result is:
```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py
black................................................(no files to check)Skipped
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py
isort................................................(no files to check)Skipped
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ 

```
There doesn't seem to be any errors from black or isort. Perhaps this is because the two files in question have been committed and so black/isort doesn't regard them as changed? (This is a guess, since I don't know what black/iisort checks for)

lkubb commented:
@dnessett $PWD being `~/bugfix-67214/salt` indicates to me you're inside the `salt` subdirectory of the repository when running the command, not in the repository root, which the paths I gave you are relative to. From pre-commit's perspective, they don't exist, so there's nothing to do. It's not about being committed, this generally works.

dnessett commented:
> @dnessett $PWD being `~/bugfix-67214/salt` indicates to me you're inside the `salt` subdirectory of the repository when running the command, not in the repository root, which the paths I gave you are relative to. From pre-commit's perspective, they don't exist, so there's nothing to do. It's not about being committed, this generally works.

I'm in the git source directory when I execute the black/isort  functions. I am reorganizing my development environment, so I cloned a copy of my github repository to a directory in my home directory in order to address this problem while the reorganization is in process.

```
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ ls
AUTHORS       CODE_OF_CONDUCT.md  doc          noxfile.py   pyproject.toml  rfcs         setup.cfg    tests
changelog     conf                LICENSE      pkg          pytest.ini      salt         setup.py     tools
CHANGELOG.md  CONTRIBUTING.rst    MANIFEST.in  POLICY.rst   README.rst      scripts      SUPPORT.rst
cicd          DEPENDENCIES.md     NOTICE       __pycache__  requirements    SECURITY.md  templates
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ pre-commit run black --files tests/minionswarm.py,tests/support/runtests.py
black................................................(no files to check)Skipped
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$ pre-commit run isort --files tests/minionswarm.py,tests/support/runtests.py
isort................................................(no files to check)Skipped
(salt) dan@Eclipse-salt:~/bugfix-67214/salt$
```

lkubb commented:
OK, sorry for the confusion!

Went back and checked, it seems I messed up and copied the wrong format. Please replace the comma with a space like this:

```bash
pre-commit run black --files tests/minionswarm.py tests/support/runtests.py
pre-commit run isort --files tests/minionswarm.py tests/support/runtests.py
```

dnessett committed:
Pre-commit black and Pre-commit isort fixed problems with tests/minionswarm.py and tests/support/runtests.py

dnessett commented:
> OK, sorry for the confusion!
> 
> Went back and checked, it seems I messed up and copied the wrong format. Please replace the comma with a space like this:
> 
> ```shell
> pre-commit run black --files tests/minionswarm.py tests/support/runtests.py
> pre-commit run isort --files tests/minionswarm.py tests/support/runtests.py
> ```

No problem. Thanks for the help.

dnessett commented:
OK. The changes made by black and isort I have pushed to https://github.com/dnessett/salt.

dwoz committed:
Fix tests/minionswarm.py import + correctness bugs

Address review feedback on #67214:

* tests/support/runtests.py: restore canonical package-qualified
  import. ``import support.paths as paths`` only resolved when the
  CWD was ``tests/`` and broke ``from tests.support.runtests import
  RUNTIME_VARS`` -- used by 50+ conftests/tests. Use
  ``from tests.support import paths`` instead. This was the root
  cause of the 393 failing Test Salt / Test Package jobs.
* tests/minionswarm.py: use ``from tests.support import runtests``
  and add a sys.path bootstrap so the script can still be executed
  directly via ``python tests/minionswarm.py`` (the original
  ``cd tests/`` reproduction from #67214 also still works).
* tests/minionswarm.py: collapse duplicated ``machine_id``
  assignment, fix invalid ``sys.exit([1])`` -> ``sys.exit(1)``.
* tests/minionswarm.py: master ``sock_dir`` pointed at
  ``var/run/salt/minion`` instead of ``var/run/salt/master``.
* tests/minionswarm.py: ``--open-mode`` was declared with
  ``default=True`` and no action, so ``--open-mode=False`` parsed
  to the string ``"False"`` (truthy). Switched to
  ``--no-open-mode`` (action="store_false") so the flag does what
  the help text claims.
* tests/minionswarm.py: typo fixes (``/tmp/srooot`` -> ``/tmp/sroot``,
  ``manaaged`` -> ``managed``, ``alt-master`` -> ``salt-master``).
